src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -220,19 +220,23 @@ internal virtual void Configure()
             // also assigns JsonPropertyInfo's JsonTypeInfo which causes SO if there are any
             // cycles in the object graph. For that reason properties cannot be added immediately.
             // This is a no-op for ReflectionJsonTypeInfo
-            LateAddProperties();
+            var propertyCache = LateAddProperties() ?? PropertyCache;
 
             DataExtensionProperty?.EnsureConfigured();
 
-            if (converter.ConverterStrategy == ConverterStrategy.Object && PropertyCache != null)
+            if (converter.ConverterStrategy == ConverterStrategy.Object && propertyCache != null)
             {
-                foreach (var jsonPropertyInfoKv in PropertyCache.List)
+                foreach (var jsonPropertyInfoKv in propertyCache.List)
                 {
                     JsonPropertyInfo jsonPropertyInfo = jsonPropertyInfoKv.Value!;
                     jsonPropertyInfo.DeclaringTypeNumberHandling = NumberHandling;
                     jsonPropertyInfo.EnsureConfigured();
                 }
 
+                // This code can be run in multiple threads therefore we assign only after all properties has been configured
```

Comment:
I was hoping we could achieve same without having to do locks but agreed it will make it harder to prevent from the regression to happen again. Will revert and lock

Reply:
actually let me close this PR and create a new one since it's effectively different change